src/simplify_qdq.cpp

```diff
@@ -0,0 +1,205 @@
+#include <migraphx/register_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+std::unordered_set<std::string> get_quantizable_op_names()
+{
+    static std::unordered_set<std::string> s = {"convolution", "dot"};
+    return s;
+}
+
+std::unordered_set<std::string> get_all_op_names()
+{
+    static auto ops = get_operators();
+    static std::unordered_set<std::string> s(ops.begin(), ops.end());
+    return s;
+}
+
+bool inputs_are_zeros(const instruction_ref& ins)
+{
+    literal zp_lit;
+    if(ins->name() == "multibroadcast" or ins->name() == "broadcast")
+        zp_lit = ins->inputs().at(0)->get_literal();
+    else
+        zp_lit = ins->get_literal();
+    std::vector<int64_t> zero_points;
+    zp_lit.visit([&](const auto zp) {
+        std::transform(
+            zp.begin(), zp.end(), std::back_inserter(zero_points), [](auto&& z) { return z; });
+    });
+    return std::all_of(
+        zero_points.begin(), zero_points.end(), [](const auto& z) { return z == 0; });
+}
+
+double get_scale(const instruction_ref& ins)
+{
+    literal scale_lit;
+    if(ins->name() == "multibroadcast" or ins->name() == "broadcast")
+        scale_lit = ins->inputs().at(0)->get_literal();
+    else
+        scale_lit = ins->get_literal();
+    std::vector<float> scales;
+    scale_lit.visit([&](auto sl) {
+        std::transform(
+            sl.begin(), sl.end(), std::back_inserter(scales), [](auto&& s) { return s; });
+    });
+    double epsilon = 1e-6;
+    if(not std::all_of(scales.begin(), scales.end(), [&](const auto& s) {
+           return std::abs(s - scales.front()) < epsilon;
+       }))
+        MIGRAPHX_THROW("Multiple scales not currently supported");
+    return scales.front();
+}
+
+instruction_ref insert_quantize_op(module& m,
+                                   instruction_ref ins,
+                                   const std::string& name,
+                                   instruction_ref x,
+                                   instruction_ref scale,
+                                   instruction_ref shift)
+{
+    auto lens = x->get_shape().lens();
+    auto scale_mb =
+        m.insert_instruction(ins, make_op("multibroadcast", {{"output_lens", lens}}), scale);
+    auto shift_mb =
+        m.insert_instruction(ins, make_op("multibroadcast", {{"output_lens", lens}}), shift);
+    return m.insert_instruction(ins, make_op(name), x, scale_mb, shift_mb);
+}
+
+struct match_find_quantizable_ops
+{
+    auto matcher() const
+    {
+        return match::name(get_all_op_names())(match::arg(0)(
+            match::name(get_quantizable_op_names())(
+                match::arg(0)(match::name("dequantizelinear")(
+                    match::arg(0)(match::name(get_all_op_names()).bind("q0")).bind("dq0"))),
+                match::arg(1)(match::name("dequantizelinear")(
+                    match::arg(0)(match::name(get_all_op_names()).bind("q1")).bind("dq1"))))
+                .bind("qop")));
```

Comment:
This matcher doesnt check for `quantizelinear`. I would write this matcher like this:

```cpp
auto dequantizelinear_op(const std::string& name, const std::string& scale)
{
    return match::name("quantizelinear")(
        match::arg(0)(match::any().bind(name)),
        match::arg(1)(match::skip_broadcasts(match::any().bind(scale))),
        match::arg(2)(match::skip_broadcasts(match::none_of(match::has_value(0))))
    );
}

auto matcher() const
{
    match::name("quantizelinear")(match::args(
        match::name(get_quantizable_op_names())(
            match::arg(0)(dequantizelinear_op("x1", "scale1")),
            match::arg(1)(dequantizelinear_op("x2", "scale2"))
        )
    ));
}
```

Reply:
I guess the above is wrong since I was assuming a quantizelinear on the output, but we will need to skip quantizelinear ops:

```cpp
MIGRAPHX_PRED_MATCHER(has_same_value, instruction_ref ins) 
{
    if (!ins->name() != "@literal")
        return false;
    bool all_same = false;
    ins->get_literal().visit([&](auto s) {
        all_same = std::all_of(s.begin()+1, s.end(), [&](const auto& scale) {
            return float_equal(scale, s.front());
        });
    });
    return all_same;
}

auto dequantizelinear_op(const std::string& name, const std::string& scale)
{
    return match::name("dequantizelinear")(
        match::arg(0)(match::skip(match::name("quantizelinear"))(mathc::any().bind(name))),
        match::arg(1)(match::skip_broadcasts(has_same_value().bind(scale))),
        match::arg(2)(match::none_of(match::has_value(0)))
    );
}

auto matcher() const
{
    match::name(get_quantizable_op_names())(
        match::arg(0)(dequantizelinear_op("x1", "scale1")),
        match::arg(1)(dequantizelinear_op("x2", "scale2"))
    );
}
```